### PR TITLE
Final modification for audit

### PR DESCRIPTION
**Pull request by jkrivine**
This PR addresses remaining issues raised by the audit:
* [FIXES] 5.1 Forwarder.provisionOf Calculation Is Wrong
  * `MangroveOffer` provides a corrected internal calculation for locked provision
  * `Forwarder` exposes the function and adds  offer's `weiBalance` to the returned value
* [FIXES] 5.2 Users Can Steal Funds From MangroveOrder
  * a difference between Mangrove's `offer_gasbase` of update offer's `gasbase` now also triggers a recomputation of `gasprice`
  * `__posthookSuccess__` of MangroveOffer now uses internal `_updateOffer`'s implementation provided by Forwarder (and Direct). This avoids successfully reposting an offer on someone else's provision.
  * as a consequence `_updateOffer` no longer takes the spurious argument `owner` (but takes extra argument `offerId` as compared to `_newOffer` which now requires an `owner` argument for Forwarder strats.
* [GAS OPTIM] using local and global configuration sent by Mangrove during `makerExecute` instead of calling `MGV.config`
* [FIXES NEW ISSUE] We found an issue w.r.t the computation of residual provision initially defined on 
`__posthookFallback__` in Forwarder strats. The issue is that a strat builder can potentially override this hook the following way: 
```solidity 
function __posthookFallback__(…) … {   
  super.__posthookFallback(…);     
  additional_behavior 
}  
```
and all gas cost incurred by the `additional_behavior` is not taken into account by the computation occurring in `super`. This would force strat developers to always call super at the end of their override which is a rather strong constraint. Instead we decided to enforce this by delaying the computation of the residual provision *after* `__posthookFallback__` in a dedicated hook called `__handleResidualProvision__` which is a noop by default.  Note the computation is unchanged simply done later.
* [MINOR] the emitted log `OrderSummary` in `MangroveOrder` has more data.

### CODE REVIEW

File: src/strategies/MangroveOrder.sol [hunk 1]
```diff
@@ -227,7 +228,8 @@ contract MangroveOrder is Forwarder, IOrderLogic {
       takerGave: res.takerGave,
       bounty: res.bounty,
       fee: res.fee,
-      restingOrderId: res.offerId
+      restingOrderId: res.offerId,
+      taker: msg.sender
```
Comment (lnist) [inline src/strategies/MangroveOrder.sol]: why this change?

File: test/lib/tokens/MintableERC20BLWithDecimals.sol [hunk 2]
```diff
@@ -11,9 +11,9 @@ contract MintableERC20BLWithDecimals is
   mapping(address => bool) admins;
   uint public _decimals; // full uint to help forge-std's stdstore
 
-  constructor(address admin, string memory name, string memory symbol, uint8 decimals) ERC20BL(name, symbol) {
+  constructor(address admin, string memory name, string memory symbol, uint8 __decimals) ERC20BL(name, symbol) {
```
Comment (lnist) [inline test/lib/tokens/MintableERC20BLWithDecimals.sol]: why this change?

**Review by lnist — COMMENTED** [review]
Are there some spurious changes in this PR or are they intended?

File: src/strategies/MangroveOrder.sol (hunk 1 above)
Comment (jkrivine) [reply src/strategies/MangroveOrder.sol]: by error

File: test/lib/tokens/MintableERC20BLWithDecimals.sol (hunk 2 above)
Comment (jkrivine) [reply test/lib/tokens/MintableERC20BLWithDecimals.sol]: to remove a warning

File: src/strategies/MangroveOrder.sol (hunk 1 above)
Comment (lnist) [reply src/strategies/MangroveOrder.sol]: should you revert the change then? :)

File: src/strategies/offer_forwarder/abstract/Forwarder.sol [hunk 3]
```diff
@@ -211,10 +216,15 @@ abstract contract Forwarder is IForwarder, MangroveOffer {
...
-        return offerId;
+        return REPOST_SUCCESS;
       } catch Error(string memory reason) {
         require(args.noRevert, reason);
-        return 0;
+        bytes32 reason_hsh = keccak256(bytes(reason));
+        if (reason_hsh == BELOW_DENSITY) {
+          return REPOST_FAILED_DUST; // offer not reposted because residual is below density
+        } else {
+          return REPOST_FAILED; // offer not reposted for other reasons (i.e lack of provision)
```
Comment (lnist) [inline src/strategies/offer_forwarder/abstract/Forwarder.sol]: Consider: do we want to return the mangrove reason instead?

Comment (lnist) [reply src/strategies/offer_forwarder/abstract/Forwarder.sol]: or in other words: why do we translate to repost_failed?

File: src/strategies/MangroveOffer.sol [hunk 4]
```diff
@@ -280,26 +283,34 @@ abstract contract MangroveOffer is AccessControlled, IOfferLogic {
...
     uint new_wants = __residualWants__(order);
-    try MGV.updateOffer(
-      order.outbound_tkn,
-      order.inbound_tkn,
-      new_wants,
-      new_gives,
-      order.offerDetail.gasreq(),
-      order.offerDetail.gasprice(),
-      order.offer.next(), // using next as pivot since this offer is off the book
+    return _updateOffer(
```
Comment (lnist) [inline src/strategies/MangroveOffer.sol]: Would it make sense to pass in order.offerDetail and order.global, order.local? This would get rid of the
```
(vars.global, vars.local) = MGV.config(address(args.outbound_tkn), address(args.inbound_tkn));
 vars.offerDetail = MGV.offerDetails(address(args.outbound_tkn), address(args.inbound_tkn), offerId);
```

in the _updateOffer function. This would save some gas for the happy path of reposting residual. For a call to updateOffer->_updateOffer from the outside we can then look these up in updateOffer.

The only drawback I see is that Direct has the same method signature and it does not need all that data.

File: src/strategies/MangroveOrder.sol (hunk 1 above)
Comment (jkrivine) [reply src/strategies/MangroveOrder.sol]: If you know how to revert this specific one yes. Otherwise I don't think this matters

File: src/strategies/offer_forwarder/abstract/Forwarder.sol (hunk 3 above)
Comment (jkrivine) [reply src/strategies/offer_forwarder/abstract/Forwarder.sol]: Good point, it requires a bit of work because the reason is a string. So we need to compare its keccak, but yes makes sense.

File: src/strategies/MangroveOffer.sol (hunk 4 above)
Comment (jkrivine) [reply src/strategies/MangroveOffer.sol]: Let's keep it aside for future gas optim work (I think there are many spots where gas can be saved)

File: src/strategies/MangroveOrder.sol (hunk 1 above)
Comment (lnist) [reply src/strategies/MangroveOrder.sol]: Simply move  taker: msg.sender to just below inbound then you have in practice reverted it.

File: src/strategies/MangroveOffer.sol [hunk 5]
```diff
@@ -260,13 +259,28 @@ abstract contract MangroveOffer is AccessControlled, IOfferLogic {
     return order.offer.gives() - order.wants;
   }
 
+  ///@notice converts Mangrove's error message string back into its original bytes32 type
+  ///@param reason the revert reason
+  function _repostStatus(string memory reason) internal pure returns (bytes32) {
```
Comment (lnist) [inline src/strategies/MangroveOffer.sol]: If I understand correctly then this method could be reduced to return bytes32(reason) except for the REPOST_FAILED case. Am I missing something?

Comment (jkrivine) [reply src/strategies/MangroveOffer.sol]: yes expect you can't cast a string to a bytes32...

**Review by lnist — APPROVED** [review]

**Review by lnist — APPROVED** [review]
